### PR TITLE
indexer JSON RPC server

### PR DESCRIPTION
**Pull request by gegaowp**
Create a skeleton for JSON RPC server to unblock module registration etc. run locally and make sure the server can start and indexer writing path can start as well

Test via pinging the local server:
```
curl --location --request POST http://127.0.0.1:3030 \
--header 'Content-Type: application/json' \
--data-raw '{ "jsonrpc":"2.0", "id":1, "method":"rpc.discover", "params":[]}'
{"jsonrpc":"2.0","result":{"openrpc":"1.2.6","info":{"title":"Sui JSON-RPC","description":"Sui JSON-RPC API for interaction with Sui Full node.","contact":{"name":"Mysten Labs","url":"https://mystenlabs.com","email":"build@mystenlabs.com"},"license":{"name":"Apache-2.0","url":"https://raw.githubusercontent.com/MystenLabs/sui/main/LICENSE"},"version":"0.0.0"},"methods":[{"name":"sui_getLatestCheckpointSequenceNumber","tags":[{"name":"Checkpoint API"}],"description":"Returns the latest checkpoint sequence number, which starts at 0 and increments by 1 each time.","params":[],"result":{"name":"i64","required":true,"schema":{"type":"integer","format":"int64"}}}],"components":{}},"id":1}%           

curl --location --request POST http://127.0.0.1:3030 \
--header 'Content-Type: application/json' \
--data-raw '{ "jsonrpc":"2.0", "id":1, "method":"sui_getLatestCheckpointSequenceNumber", "params":[]}'
{"jsonrpc":"2.0","result":523,"id":1}%      
```

### CODE REVIEW

File: crates/sui-indexer/src/lib.rs [hunk 1]
```diff
@@ -66,3 +73,24 @@ pub fn get_pg_pool_connection(
         ))
     })
 }
+
+pub async fn build_json_rpc_server(
+    prometheus_registry: &Registry,
+) -> Result<ServerHandle, IndexerError> {
+    let builder = ExtendedJsonRpcServerBuilder::new(FAKE_PKG_VERSION, prometheus_registry)
```
Comment (patrickkuo) [inline crates/sui-indexer/src/lib.rs:80]: why do we need the new crate and `ExtendedJsonRpcServerBuilder`? 

we should be able to re-use `JsonRpcServerBuilder` here directly?

**Review by patrickkuo — CHANGES_REQUESTED** [review]

File: crates/sui-indexer/src/lib.rs (hunk 1 above)
Comment (gegaowp) [reply crates/sui-indexer/src/lib.rs:80]: you are right, I was originally thinking that we will have similar directory & file names for APIs, so it is cleaner to separate them, but they do not have to be a different crate, I can add a new sub-dir for that.

**Comment by gegaowp** [comment]
@patrickkuo revised the PR and reused the same crate, when I tried to ping the server, I saw this, did I miss anything
```
curl -X POST -H "Content-Type: application/json" -d '{"method": "rpc.discover"}' 127.0.0.1:3030
curl: (7) Failed to connect to 127.0.0.1 port 3030 after 2 ms: Couldn't connect to server
```

File: crates/sui-indexer/src/main.rs [hunk 2]
```diff
@@ -26,6 +26,25 @@ async fn main() -> Result<(), IndexerError> {
...
+        .unwrap(),
+    );
+
+    let json_rpc_prometheus_registry = registry_service.default_registry();
+    let pg_connection_pool = new_pg_connection_pool(indexer_config.db_url.clone()).await?;
+    let handle = build_json_rpc_server(&json_rpc_prometheus_registry, pg_connection_pool.clone())
+        .await
+        .expect("Json rpc server should not run into errors upon start.");
+    // let JSON RPC server run forever.
+    tokio::spawn(handle.stopped());
```
Comment (patrickkuo) [inline crates/sui-indexer/src/main.rs:46]: we can use `spawn_monitored_task` instead of Tokyo::spawn

**Review by patrickkuo — APPROVED** [review]
LGTM!

File: crates/sui-indexer/src/apis/checkpoint_api.rs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+pub struct CheckpointApiImpl {
+    pg_connection_pool: Arc<PgConnectionPool>,
+}
+
+impl CheckpointApiImpl {
+    pub fn new(pg_connection_pool: Arc<PgConnectionPool>) -> Self {
+        Self { pg_connection_pool }
+    }
+
+    async fn get_total_checkpoint(&self) -> Result<i64, IndexerError> {
```
Comment (patrickkuo) [inline crates/sui-indexer/src/apis/checkpoint_api.rs:26]: actually, isn't this same as `sui_getLatestCheckpointSequenceNumber` ?

**Review by patrickkuo — CHANGES_REQUESTED** [review]
`get_total_checkpoint` is same as `sui_getLatestCheckpointSequenceNumber`? as checkpoints are sequential

**Review by patrickkuo — APPROVED** [review]

**Pull request merged**